File: registry/src/domain.ts
/**
 * Domain separation, record hashing and signing input.
 *
 * docs/spec/REGISTRY.md, "Canonical Serialisation and Signing":
 *
 *   signing_input = "VayuWeb-Registry-Record-v1" || 0x00 || uint8(suite) || det_cbor(core)
 *   record_hash   = BLAKE2b-256("VayuWeb-Registry-Hash-v1" || 0x00 || det_cbor(full))
 *
 * `core` is the record map with `sig` and `coSig` removed; `full` is the complete map
 * including them.
 *
 * The prefixes are what stop a signature made over one VayuWeb structure being replayed as
 * another, and stop a signing input being read as a hash preimage. They are therefore
 * consensus-critical constants: an implementation that gets a single byte wrong produces
 * signatures no other implementation will accept, which is a silent fork rather than a visible
 * error. The byte lengths are asserted at module load and pinned again in the tests, because
 * this is a value that must never drift, and a specification prose error about it has already
 * happened once — REGISTRY.md described these as 23 and 21 bytes when the literal strings are
 * 26 and 24. The literal strings in the normative code block are authoritative; the prose
 * count was the error.
 */

import { blake2b } from '@noble/hashes/blake2';

import { encode, type CborMap, type CborValue } from './cbor.ts';

/** Domain-separation prefix for the Ed25519 signing input over a registry record. */
export const RECORD_SIGNING_PREFIX = 'VayuWeb-Registry-Record-v1';

/** Domain-separation prefix for the BLAKE2b-256 record hash. */
export const RECORD_HASH_PREFIX = 'VayuWeb-Registry-Hash-v1';

/** Length of a record hash in bytes. BLAKE2b-256, matching Hypercore's own primitive. */
export const RECORD_HASH_LENGTH = 32;

/** The all-zero hash used as `prevHash` when `seq` is 0. */
export const ZERO_HASH: Uint8Array = new Uint8Array(RECORD_HASH_LENGTH);

/** Fields excluded from the signing input, because they carry the signatures themselves. */
const SIGNATURE_FIELDS = ['sig', 'coSig'] as const;

const asciiPrefix = (literal: string): Uint8Array => {
  const bytes = new TextEncoder().encode(literal);
  for (const b of bytes) {
    if (b < 0x20 || b > 0x7e) {
      throw new Error(`domain prefix must be printable ASCII: ${literal}`);
    }
  }
  return bytes;
};

const SIGNING_PREFIX_BYTES = asciiPrefix(RECORD_SIGNING_PREFIX);
const HASH_PREFIX_BYTES = asciiPrefix(RECORD_HASH_PREFIX);

// Fail at load rather than at first divergence. A wrong prefix length is not something to
// discover from a peer refusing your records.
if (SIGNING_PREFIX_BYTES.length !== 26) {
  throw new Error(`signing prefix must be 26 bytes, got ${SIGNING_PREFIX_BYTES.length}`);
}
if (HASH_PREFIX_BYTES.length !== 24) {
  throw new Error(`hash prefix must be 24 bytes, got ${HASH_PREFIX_BYTES.length}`);
}

// The two prefixes must not share a prefix relationship, or a crafted body could make one
// domain's input parse as another's. Distinct lengths plus distinct content is sufficient here,
// but the check is cheap and survives someone editing the literals later.
if (
  RECORD_SIGNING_PREFIX.startsWith(RECORD_HASH_PREFIX) ||
  RECORD_HASH_PREFIX.startsWith(RECORD_SIGNING_PREFIX)
) {
  throw new Error('domain prefixes must not be prefixes of one another');
}

/** Concatenate a domain prefix, its 0x00 separator, optional extra bytes, and a body. */
function withDomain(
  prefix: Uint8Array,
  body: Uint8Array,
  extra: readonly number[] = [],
): Uint8Array {
  const out = new Uint8Array(prefix.length + 1 + extra.length + body.length);
  out.set(prefix, 0);
  out[prefix.length] = 0x00;
  out.set(extra, prefix.length + 1);
  out.set(body, prefix.length + 1 + extra.length);
  return out;
}

/**
 * Return a copy of `record` with the signature fields removed — the `core` map.
 *
 * A copy is returned rather than mutating in place: the caller's record is frequently the
 * object about to be stored or replicated, and silently stripping its signature would be a
 * spectacular action at a distance.
 */
export function core(record: CborMap): CborMap {
  const stripped: CborMap = new Map();
  for (const [key, value] of record) {
    if (typeof key === 'string' && (SIGNATURE_FIELDS as readonly string[]).includes(key)) {
      continue;
    }
    stripped.set(key, value);
  }
  return stripped;
}

/**
 * The exact bytes an Ed25519 signature is made over.
 *
 * Unknown fields are deliberately retained: REGISTRY.md requires that a record carrying fields
 * this version does not understand still verifies downstream, so the signing input covers
 * whatever the author actually signed rather than only the fields we recognise.
 */
export function signingInput(record: CborMap): Uint8Array {
  // CRYPTO-AGILITY.md 4.3 puts the suite identifier inside the domain-separated prefix, so a
  // signature made under one suite cannot be replayed as another. Belt and braces: `suite` is
  // also a field inside `core` and so already covered. The requirement is kept because the cost
  // is one byte and the failure it prevents — a cross-suite replay during the one migration this
  // protocol gets — is not recoverable.
  //
  // A missing or malformed `suite` throws rather than defaulting. This function is called to
  // BUILD records as well as to check them, and a builder that omitted the field would otherwise
  // produce a signature over the pre-agility input — valid-looking bytes that no verifier
  // implementing 4.3 will accept, discovered only when a peer refuses them.
  const suite = record.get('suite');
  if (typeof suite !== 'number' || !Number.isInteger(suite) || suite < 0 || suite > 255) {
    throw new Error(
      'signingInput: a record needs an integer `suite` in 0..255 (CRYPTO-AGILITY 4.3)',
    );
  }
  return withDomain(SIGNING_PREFIX_BYTES, encode(core(record)), [suite]);
}

/**
 * The record hash: BLAKE2b-256 over the domain-separated bytes, hashed as they arrived.
 *
 * It covers the full record including `sig` and `coSig`, because the hash identifies the record as
 * it exists on the wire — which is what `prevHash` chains to and what the convergence tie-break
 * compares.
 *
 * REGISTRY.md: "A peer MUST NOT re-serialise a record it did not author: received bytes are
 * stored and replicated verbatim". Decoding and re-encoding to compute a hash would defeat that,
 * so this takes bytes and there is no longer a sibling that takes a map.
 *
 * **There was one, called `recordHash`, and the shorter name did the forbidden thing.** Anyone
 * reaching for "hash a record" found it first and got a hash that is right only if the record
 * re-encodes byte-identically — true here, but only because `cbor.ts` refuses non-canonical input
 * on decode, which is a guarantee made in another module and stated in neither. Nothing authored a
 * record and needed its hash from the map, so the trap went rather than being documented.
 */
export function recordHashFromBytes(recordBytes: Uint8Array): Uint8Array {
  return blake2b(withDomain(HASH_PREFIX_BYTES, recordBytes), { dkLen: RECORD_HASH_LENGTH });
}

/** Constant-time equality for hashes and other fixed-length secrets-adjacent values. */
export function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i] ^ b[i];
  return diff === 0;
}

/** True when `hash` is the all-zero hash required of a `REGISTER` record's `prevHash`. */
export function isZeroHash(hash: Uint8Array): boolean {
  return bytesEqual(hash, ZERO_HASH);
}

export type { CborMap, CborValue };

File: client/src/domain.rs
//! Domain separation, record hashing and signing input.
//!
//! The client-side counterpart of `registry/src/domain.ts`. REGISTRY.md fixes the two inputs
//! byte for byte:
//!
//! ```text
//! signing_input = "VayuWeb-Registry-Record-v1" || 0x00 || uint8(suite) || det_cbor(core)
//! record_hash   = BLAKE2b-256("VayuWeb-Registry-Hash-v1" || 0x00 || det_cbor(full))
//! ```
//!
//! `core` is the record map with `sig` and `coSig` removed; `full` is the complete map. The
//! prefixes are what stop a signature made over one VayuWeb structure being replayed as
//! another, so they are consensus-critical constants: an implementation one byte wrong
//! produces signatures no other implementation accepts, which is a silent fork rather than a
//! visible error. Their lengths are asserted where they are built and pinned again by tests,
//! because a specification prose error about exactly this has happened before — REGISTRY.md
//! once described them as 23 and 21 bytes when the literals are 26 and 24.
//!
//! What proves this port agrees with the reference implementation is again bytes rather than
//! prose: the golden fixtures under `conformance/` are built here, and the registry's verifier
//! recomputes `record_hash` over those same bytes and refuses any disagreement.

use blake2::digest::{Update, VariableOutput};
use blake2::Blake2bVar;

use crate::cbor::{self, Key, Value};

/// Domain-separation prefix for the Ed25519 signing input over a registry record.
pub const RECORD_SIGNING_PREFIX: &str = "VayuWeb-Registry-Record-v1";

/// Domain-separation prefix for the BLAKE2b-256 record hash.
pub const RECORD_HASH_PREFIX: &str = "VayuWeb-Registry-Hash-v1";

/// Length of a record hash in bytes. BLAKE2b-256, matching Hypercore's own primitive.
pub const RECORD_HASH_LENGTH: usize = 32;

/// Fields excluded from the signing input, because they carry the signatures themselves.
const SIGNATURE_FIELDS: [&str; 2] = ["sig", "coSig"];

fn ascii_prefix(literal: &str) -> Vec<u8> {
    let bytes = literal.as_bytes();
    for &b in bytes {
        if !(0x20..=0x7e).contains(&b) {
            panic!("domain prefix must be printable ASCII: {literal}");
        }
    }
    bytes.to_vec()
}

fn signing_prefix_bytes() -> Vec<u8> {
    let bytes = ascii_prefix(RECORD_SIGNING_PREFIX);
    assert_eq!(bytes.len(), 26, "signing prefix must be 26 bytes");
    bytes
}

fn hash_prefix_bytes() -> Vec<u8> {
    let bytes = ascii_prefix(RECORD_HASH_PREFIX);
    assert_eq!(bytes.len(), 24, "hash prefix must be 24 bytes");
    bytes
}

/// Concatenate a domain prefix, its 0x00 separator, optional extra bytes, and a body.
fn with_domain(prefix: &[u8], body: &[u8], extra: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(prefix.len() + 1 + extra.len() + body.len());
    out.extend_from_slice(prefix);
    out.push(0x00);
    out.extend_from_slice(extra);
    out.extend_from_slice(body);
    out
}

/// The record map without its signature fields, ready for canonical encoding.
///
/// Removing rather than omitting is the point: a caller that builds `core` separately from
/// `full` can build them differently, which is how a signature ends up over bytes the record
/// no longer matches. One map, stripped twice-verified, is one fact.
pub fn core_of(record: &[(Key, Value)]) -> Result<Vec<(Key, Value)>, cbor::CborError> {
    let mut core = Vec::with_capacity(record.len());
    for (key, value) in record {
        let Key::Text(name) = key else {
            return Err(cbor::CborError::BadMapKey);
        };
        if SIGNATURE_FIELDS.contains(&name.as_str()) {
            continue;
        }
        core.push((key.clone(), value.clone()));
    }
    Ok(core)
}

/// Canonical CBOR of the record map with `sig` and `coSig` removed.
pub fn encode_core(record: &[(Key, Value)]) -> Result<Vec<u8>, cbor::CborError> {
    cbor::encode(&Value::Map(core_of(record)?))
}

/// `signing_input`: what the owner's Ed25519 signature is computed over.
pub fn signing_input(core_cbor: &[u8], suite: u8) -> Vec<u8> {
    with_domain(&signing_prefix_bytes(), core_cbor, &[suite])
}

/// `record_hash`: BLAKE2b-256 over the domain-separated full record bytes.
///
/// BLAKE2b's variable digest length is part of the function's parameterisation, not a
/// truncation of a longer digest — the same reason the reference implementation passes a
/// digest length rather than hashing wide and cutting.
pub fn record_hash_from_bytes(record_bytes: &[u8]) -> [u8; RECORD_HASH_LENGTH] {
    let input = with_domain(&hash_prefix_bytes(), record_bytes, &[]);
    let mut hasher = Blake2bVar::new(RECORD_HASH_LENGTH).expect("32 is a valid BLAKE2b width");
    Update::update(&mut hasher, &input);
    let mut out = [0u8; RECORD_HASH_LENGTH];
    hasher
        .finalize_variable(&mut out)
        .expect("exact output size");
    out
}

/// Constant-time equality for hashes and other fixed-length values.
pub fn bytes_equal(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hex(bytes: &[u8]) -> String {
        bytes.iter().map(|b| format!("{b:02x}")).collect()
    }

    #[test]
    fn prefixes_are_the_specification_literals_with_the_specification_lengths() {
        // Asserted at build time too; asserted here so a test failure reads as the defect it
        // is rather than a panic inside hashing.
        assert_eq!(RECORD_SIGNING_PREFIX.len(), 26);
        assert_eq!(RECORD_HASH_PREFIX.len(), 24);
        // Neither prefix is a prefix of the other, so no input is well-formed under both
        // domains at once.
        assert!(!RECORD_HASH_PREFIX.starts_with(RECORD_SIGNING_PREFIX));
        assert!(!RECORD_SIGNING_PREFIX.starts_with(RECORD_HASH_PREFIX));
        // And the exact bytes are pinned, because these ARE the wire-visible constants.
        assert_eq!(
            hex(&ascii_prefix(RECORD_SIGNING_PREFIX)),
            "566179755765622d52656769737472792d5265636f72642d7631"
        );
        assert_eq!(
            hex(&ascii_prefix(RECORD_HASH_PREFIX)),
            "566179755765622d52656769737472792d486173682d7631"
        );
    }

    #[test]
    fn signing_input_shape_is_prefix_zero_suite_body() {
        let core = vec![0xde, 0xad, 0xbe, 0xef];
        let input = signing_input(&core, 1);
        assert_eq!(input.len(), 26 + 1 + 1 + core.len());
        assert_eq!(&input[..26], RECORD_SIGNING_PREFIX.as_bytes());
        assert_eq!(input[26], 0x00);
        assert_eq!(input[27], 1);
        assert_eq!(&input[28..], &core);
        // The suite byte is part of the domain: the same core under a different suite is a
        // different signing input.
        assert_ne!(signing_input(&core, 1), signing_input(&core, 2));
    }

    #[test]
    fn core_of_strips_only_signature_fields_and_keeps_everything_else() {
        let record = vec![
            (Key::Text("version".into()), Value::UInt(1)),
            (Key::Text("op".into()), Value::Text("REGISTER".into())),
            (Key::Text("sig".into()), Value::Bytes(vec![0xaa; 64])),
            (Key::Text("coSig".into()), Value::Bytes(vec![0xbb; 64])),
        ];
        let core = core_of(&record).expect("strips");
        assert_eq!(core.len(), 2);
        assert_eq!(core[0].0, Key::Text("version".into()));
        assert_eq!(core[1].0, Key::Text("op".into()));
        // Idempotent: stripping a core changes nothing.
        assert_eq!(core_of(&core).expect("strips"), core);
        // A record whose sig field is absent strips nothing and encodes unchanged.
        let unsigned = vec![(Key::Text("op".into()), Value::Text("RENEW".into()))];
        assert_eq!(
            encode_core(&unsigned).expect("encodes"),
            cbor::encode(&Value::Map(unsigned.clone())).expect("encodes")
        );
    }

    #[test]
    fn record_hash_is_blake2b_256_over_domain_separated_bytes() {
        let bytes = vec![0x01, 0x02, 0x03];
        let hash = record_hash_from_bytes(&bytes);
        assert_eq!(hash.len(), 32);
        // Deterministic, and distinct for distinct inputs including the empty one.
        assert_eq!(hash, record_hash_from_bytes(&bytes));
        assert_ne!(hash, record_hash_from_bytes(&[]));
        assert_ne!(record_hash_from_bytes(&[]), [0u8; 32]);
        // Distinct from the signing input of the same-length prefix: the two domains must not
        // share outputs even by accident, and a 40-byte core makes both sides long enough to
        // compare against the full 32-byte hash.
        let long_core = vec![0x07u8; 40];
        assert_ne!(&hash[..], &signing_input(&long_core, 1)[..32]);
    }

    #[test]
    fn bytes_equal_is_length_safe() {
        assert!(bytes_equal(b"abcd", b"abcd"));
        assert!(!bytes_equal(b"abc", b"abcd"));
        assert!(!bytes_equal(b"abcd", b"abce"));
    }
}
